lib/dialyzer/test/incremental_SUITE.erl

```diff
@@ -0,0 +1,881 @@
+    compile:file(filename:join(DataDir, "fix.erl"),
+                 [{outdir, PrivDir},
+                  debug_info]).
+
+%% There is one warning per module which will be reported when that module is re-analyzed
+%% We expect a module to be re-analyzed when it has changed, or when one of its transitive
+%% dependencies have changed.
+expected_modules_reanalyzed_incremental(ChangedMods) ->
+    DependsOn = fun depends_on/1,
+    ordsets:from_list([Mod || C <- ChangedMods,
+                              Mod <- [C|depends_on_transitively(C, DependsOn)]]).
+
+change_module(DataDir, PrivDir, Mod) ->
+    compile:file(filename:join(DataDir, atom_to_list(Mod)++".erl"),
+        [{outdir, PrivDir}, debug_info, {d, Mod, erlang:monotonic_time()}]).
+
+delete_module_beam_file(PrivDir, Mod) ->
+    ModuleBeamPath = PrivDir ++ atom_to_list(Mod) ++ ".beam",
+    ok = file:delete(ModuleBeamPath).
+
+%% Common utility functions
+
+replace_location(Warnings) ->
+    [{Type, loc, Warning} || {Type, _Loc, Warning} <- Warnings].
+
+warning_for_module(Mod) -> {warn_contract_types,loc,
+                       {invalid_contract,
+                           [Mod,wrong,1,"(integer()) -> integer()"]}}.
+
+compile_all(DataDir, PrivDir) ->
+    [{ok, _} = compile:file(File, [{outdir, PrivDir}, debug_info]) || File <- filelib:wildcard(DataDir ++ "*.erl")].
+
+%% This describes the dependencies between modules m1->m6 in
+%% incremental_SUITE_data/ If you change thos deps update here.
+%% Don't add cycles (yet) or at least without changing depends on transitively
+depends_on(m1) -> [];
+depends_on(m2) -> [m1];
+depends_on(m3) -> [];
+depends_on(m4) -> [m2,m3];
+depends_on(m5) -> [m4];
+depends_on(m6) -> [m4].
+
+depends_on(X, IsInPlt) ->
+    [D || D <- depends_on(X), IsInPlt(D)].
+
+depends_on_transitively(M, DependsOn) ->
+    Val = DependsOn(M),
+    Val ++ [Dep || V <- Val, Dep <- [V|depends_on_transitively(V, DependsOn)]].
+
+all_subsets([X|Rest]) ->
+    [Res || SubSet <- all_subsets(Rest), Res <- [[X|SubSet], SubSet]];
+all_subsets([]) -> [[]].
+
+all_mods() -> [m1,m2,m3,m4,m5,m6].
+
+run_dialyzer(Analysis, Files, Opts) ->
+    dialyzer:run([{analysis_type, Analysis},
+		  {files, Files},
+		  {from, byte_code}|
+		  Opts]).
+
+run_dialyzer_for_modules_analyzed(Analysis, Files, Opts) ->
+    dialyzer:run_report_modules_analyzed([{analysis_type, Analysis},
+		  {files, Files},
+		  {from, byte_code}|
+		  Opts]).
+
+default_apps_config_xdg(Config) ->
+    TestHome = filename:join(?config(priv_dir, Config), ?FUNCTION_NAME),
+
+    %% We change the $HOME of the emulator to run this test
+    HomeEnv = case os:type() of
+                  {win32, _} ->
+                      [Drive | Path] = filename:split(TestHome),
+                      [{"APPDATA", filename:join(TestHome,"AppData")},
+                       {"HOMEDRIVE", Drive}, {"HOMEPATH", Path}];
+                  _ ->
+                      [{"HOME", TestHome}]
+              end,
+
+    ConfigSubDirs =
```

Comment:
How about starting the peer node earlier and then doing `erpc:call(Node, filename, basedir, ...` to figure out where to put the config file?

Reply:
I am not entirely sure what you mean. Is your proposal to query the `Node` for the appropriate path?

Reply:
I would say that this test case tests "if I put a config file in the XDG directory it should be used", i.e. it is not important to test where the XDG directory is, but rather that the config file is used if places there.

In any case, this test case will fail on windows as they have yet another way to write the XDG path.

```erlang
5> filename:basedir(user_config, "erlang", #{ os => windows }). 
"/home/lukas/Roaming/erlang"
```

This raises another question, what happens if I set `DIALYZER_CONFIG` and have an XDG config? Does that behavior need a test?

Reply:
Ah, yes, good spot!

Regarding `DIALYZER_CONFIG` and XDG, `default_apps_config_env_var` covers this implicitly, but the semantics are that if `DIALYZER_CONFIG` exists, it is used, and if not, we fallback to the XDG location.

Reply:
>  `default_apps_config_env_var` covers this implicitly

I must be missing something as it is not obvious to me that this is the case.

Reply:
I agree this should be clearer. I’ve got another test that demonstrates this explicitly. I’m away for a few days at the moment, but I’ll push it when I get back.